Guild Resource Support

CoreyShupe opened the pull request:
Addresses https://github.com/princesslana/ErisCasper.java/issues/126

CoreyShupe committed:
Guild resources
Addresses https://github.com/princesslana/ErisCasper.java/issues/126

CoreyShupe committed:
Merge remote-tracking branch 'origin/master'

CoreyShupe committed:
spotless

ianagbip1oti commented on src/main/java/com/github/princesslana/eriscasper/rest/GuildRoute.java:
```diff
@@ -11,22 +40,323 @@ private GuildRoute(Snowflake id) {
+    return Route.get(
+        path("/bans/%s" + userId.unwrap()),
+        (response) ->
+            response.code() == 404
```
We shouldn't be dealing with 404 here. Most likely this throws an exception when the response is received anyway.

CoreyShupe replied:
No, it should be a clean 404 according to documentation.

ianagbip1oti replied:
Anything other than a 2xx response is treated as an error at the time we get the http response.

Handling this in a way other than an exception is a job for layers of code above this.

CoreyShupe replied:
We should handle it above then because we are expecting a 404 when there's no ban set for a user.

ianagbip1oti replied:
Sure, we're expecting 404 in a number of places I imagine (e.g., if they try an get a guild with an id that doesn't exist)

CoreyShupe replied:
Alright, makes sense, they are able to get all the bans, so it might not be important to watch for it here.

ianagbip1oti commented on src/main/java/com/github/princesslana/eriscasper/rest/GuildRoute.java:
```diff
@@ -11,22 +40,323 @@ private GuildRoute(Snowflake id) {
+    return Route.get(
+        path("/vanity-url"),
+        response ->
+            response.code() == 204
```
I don't see this 204 code in the docs. This has been tested?

CoreyShupe replied:
I forget why I did that. I'm sure I saw something, but we can change it to a clean change into PartialInvite.

ianagbip1oti commented on src/main/java/com/github/princesslana/eriscasper/rest/GuildRoute.java:
```diff
@@ -40,4 +370,12 @@ public static GuildRoute on(Snowflake id) {
+   * @see <a href="https://discordapp.com/developers/docs/resources/guild#create-guild">
+   *     https://discordapp.com/developers/docs/resources/guild#create-guild</a>
+   */
+  public static Route<GuildCreateRequest, Guild> createNewGuild() {
```
s/createNewGuild/createGuild/ (match discord name)

CoreyShupe replied:
will change

CoreyShupe committed:
Change nullable to RxJava

ianagbip1oti commented on src/main/java/com/github/princesslana/eriscasper/rest/guild/AddGuildMemberRequest.java:
```diff
@@ -0,0 +1,23 @@
+ *     https://discordapp.com/developers/docs/resources/guild#add-guild-member-json-params</a>
+ */
+@Value.Immutable
+public interface AddGuildMemberRequest {
```
All these fields should be optional?

CoreyShupe replied:
didn't see that, will fix

ianagbip1oti commented on src/main/java/com/github/princesslana/eriscasper/rest/guild/BeginGuildPruneRequest.java:
```diff
@@ -0,0 +1,22 @@
+    return new QueryStringBuilder().add("days", getDays().toString()).build();
+  }
+
+  static BeginGuildPruneRequest from(int days) {
```
I believe `of` would be a more consistent name used by immutables.

Do we have these methods in these classes elsewhere?

CoreyShupe replied:
in a few of them which only require one param or simple calls, similar to what was done with CreateMessage.

ianagbip1oti replied:
Cool. I thought we had put those methods elsewhere. `of` is the more consistent name.

CoreyShupe replied:
Gotcha, I'll change it.

ianagbip1oti commented on src/main/java/com/github/princesslana/eriscasper/rest/guild/CreateGuildBanRequest.java:
```diff
@@ -0,0 +1,23 @@
+  default String toQueryString() {
+    return new QueryStringBuilder()
+        .add("reason", getReason())
+        .add("delete_message_days", getDeleteMessageDays().toString())
```
Shouldn't this make use of addLong?

CoreyShupe replied:
it's non optional

ianagbip1oti replied:
You could call with Optional.of(getDeleteMessageDays()) or add an `addLong(String, Long)` method to QueryStringBuilder.

CoreyShupe replied:
Alright, I'll do that.

ianagbip1oti commented on src/main/java/com/github/princesslana/eriscasper/rest/guild/CreateGuildBanRequest.java:
```diff
@@ -0,0 +1,23 @@
+ */
+@Value.Immutable
+public interface CreateGuildBanRequest {
+  Integer getDeleteMessageDays();
```
We've used Long for integer fields elsewhere

CoreyShupe replied:
yeah I changed it

ianagbip1oti commented on src/main/java/com/github/princesslana/eriscasper/rest/guild/ListGuildMemberRequest.java:
```diff
@@ -0,0 +1,42 @@
+    return withLimitAfter(null, id);
+  }
+
+  static ListGuildMemberRequest withLimitAfter(@Nullable Long limit, @Nullable Snowflake after) {
```
By the time we get to adding methods for all possible combos I think we gain not much. It would also be nice to avoid nullable fields in public methods.

CoreyShupe replied:
Alright, I'll remove them then, you're right.

ianagbip1oti commented on src/main/java/com/github/princesslana/eriscasper/rest/guild/ModifyCurrentNickRequest.java:
```diff
@@ -0,0 +1,17 @@
+public interface ModifyCurrentNickRequest {
+  String getNick();
+
+  static ModifyCurrentNickRequest withNick(String nick) {
```
`ofNick`.

It does sound a little funky, but we use of for construction of classes elsewhere. `withXxx` methods are generated by immutables and have a different use.

CoreyShupe replied:
gotcha

CoreyShupe committed:
Small updates conforming to comments.
Change from optional requests to defined requests.
Removed nullable from ListGuildMembers
Naming conformity.

CoreyShupe committed:
QueryStringBuilder update to accept Long.

CoreyShupe committed:
Type consistency fix.

ianagbip1oti commented on src/main/java/com/github/princesslana/eriscasper/rest/guild/CreateGuildRoleRequest.java:
```diff
@@ -0,0 +1,30 @@
+
+  Optional<Boolean> isMentionable();
+
+  static CreateGuildRoleRequest newRole() {
```
Would would we expect this to be used for?

CoreyShupe replied:
Yeah, I mean just making a standard empty new role to work with later, works the same way in the client, just creates a default role to work with and position. If you want I can remove it.

ianagbip1oti replied:
Do we think people would create a new role and update it with a bot? Or just create the role with the details they need in the create request?

CoreyShupe replied:
You're right, I'll remove it. Most likely situation is creating with information.

ianagbip1oti commented on src/main/java/com/github/princesslana/eriscasper/rest/guild/GetGuildPruneCountRequest.java:
```diff
@@ -0,0 +1,22 @@
+    return new QueryStringBuilder().add("days", getDays().toString()).build();
+  }
+
+  static GetGuildPruneCountRequest from(int days) {
```
s/from/of

CoreyShupe replied:
Gotcha

ianagbip1oti commented on src/main/java/com/github/princesslana/eriscasper/rest/guild/CreateGuildRoleRequest.java:
```diff
@@ -0,0 +1,30 @@
+    return ImmutableCreateGuildRoleRequest.builder().build();
+  }
+
+  static CreateGuildRoleRequest of(String name) {
```
Is naming this `ofName` more consistent with other `of` methods we have?

CoreyShupe replied:
Not sure, we kind of switch, `of` not `ofDays` `ofNick` not `of`, so I mean it kind of goes back and forth.

CoreyShupe replied:
We should decide on a format and be consistent with it, it's mostly my fault for switching in between, just didn't know of the right format to use.

ianagbip1oti replied:
I thought we might.

For me I think we should stick with `ofXxx` going forward, because using just `of` breaks down if there's multiple of's I might want that take a String.

CoreyShupe replied:
Alright, I agree, should I change all past `of`s, or just from now on be consistent with ofXxx.

ianagbip1oti replied:
Nah, we don't need to change past ones yet. Let's be consistent going forward, and if we're woking on a class or functionality that has an `of`, we can change it then.

CoreyShupe replied:
Alright, sounds good, I'll change all of GuildRoute's `of`s just to keep this PR consistent and think about it going forward.

CoreyShupe committed:
Naming consistency and useless method removal.

ianagbip1oti commented on src/main/java/com/github/princesslana/eriscasper/rest/guild/GuildChannelCreateRequest.java:
```diff
@@ -0,0 +1,39 @@
+    return ImmutableGuildChannelCreateRequest.builder().name(name).build();
+  }
+
+  static GuildChannelCreateRequest ofNameAndCategory(String name, Snowflake category) {
```
This is where of will start to be troublesome, cos methods names with And in them are usually bad.

But we'll keep this for now and see how it goes.

CoreyShupe replied:
what makes them bad?
